File: scripts/verify-ordinary-differential-equations-independent.py
"""Complementary exact/high-precision checks against final production exports."""
import contextlib
import hashlib
import io
import json
import math
import platform
from datetime import datetime, timezone
from fractions import Fraction as F
from pathlib import Path

import mpmath as mp
import numpy as np
import scipy
from scipy.integrate import quad, solve_ivp

directory = Path('scratch/ordinary-differential-equations-independent')
data = json.loads((directory / 'fixtures.json').read_text(encoding='utf-8'))
mp.mp.dps = 420
counts = {}


def checked(kind):
    counts[kind] = counts.get(kind, 0) + 1


def close(actual, reference, relative=2e-12, absolute=0):
    rounded = float(reference)
    assert math.isfinite(actual) and math.isfinite(rounded)
    assert abs(actual - rounded) <= absolute + relative * abs(rounded), (actual, rounded)


for source in data['sources']:
    assert hashlib.sha256(Path(source['path']).read_bytes()).hexdigest() == source['sha256']

for row in data['fixtures']['cooling']:
    initial, equilibrium, time, rate = (mp.mpf(row[key]) for key in ('initial', 'equilibrium', 'time', 'rate'))
    exact = equilibrium + (initial - equilibrium) * mp.exp(-rate * time)
    close(row['result']['state'], exact)
    close(row['result']['derivative'], -rate * (exact - equilibrium), absolute=3e-10)
    checked('highPrecisionCooling')

for row in data['fixtures']['logistic']:
    initial, capacity, time, rate = (mp.mpf(row[key]) for key in ('initial', 'capacity', 'time', 'rate'))
    exact = capacity * initial / (initial + (capacity - initial) * mp.exp(-rate * time))
    close(row['result']['state'], exact, relative=3e-12, absolute=math.ulp(0.0))
    checked('highPrecisionLogistic')

for row in data['fixtures']['matrices']:
    reference = mp.expm(mp.matrix(row['matrix']) * mp.mpf(row['time']))
    for i in range(2):
        for j in range(2):
            close(row['result'][i][j], reference[i, j], absolute=2e-12)
    checked('independentMatrixExponential')

for row in data['fixtures']['forcing']:
    time, switch = row['time'], row['switchTime']
    first = quad(lambda s: 0.1 * row['firstPower'] * math.exp(-0.2 * (time - s)), 0, min(time, switch), epsabs=1e-12)[0]
    second = quad(lambda s: 0.1 * row['secondPower'] * math.exp(-0.2 * (time - s)), min(time, switch), time, epsabs=1e-12)[0]
    close(row['result']['state'], row['initial'] * math.exp(-0.2 * time) + first + second)
    close(row['result']['firstContribution'], first)
    close(row['result']['secondContribution'], second)
    checked('forcingQuadrature')


def exact_step(method, time, state, width):
    # Rational Butcher table evaluation with dyadic input, separately from JS control flow.
    table = {
        'euler': ([F(0)], [[]], [F(1)]),
        'midpoint': ([F(0), F(1, 2)], [[], [F(1, 2)]], [F(0), F(1)]),
        'rk4': ([F(0), F(1, 2), F(1, 2), F(1)], [[], [F(1, 2)], [F(0), F(1, 2)], [F(0), F(0), F(1)]], [F(1, 6), F(1, 3), F(1, 3), F(1, 6)]),
    }
    c, a, b = table[method]
    slopes, stages = [], []
    for index, offset in enumerate(c):
        stage_time = time + offset * width
        stage_state = state + width * sum((coefficient * slopes[j] for j, coefficient in enumerate(a[index])), F(0))
        slope = stage_time - 2 * stage_state
        slopes.append(slope)
        stages.append((stage_time, stage_state, slope))
    return state + width * sum((weight * slope for weight, slope in zip(b, slopes)), F(0)), stages


for row in data['fixtures']['stages']:
    exact, stages = exact_step(row['method'], F(row['time']), F(row['state']), F(row['step']))
    close(row['result']['state'], exact, absolute=2e-15)
    for actual, expected in zip(row['result']['stages'], stages):
        for key, value in zip(('time', 'state', 'slope'), expected):
            assert actual[key] == float(value), (key, actual, expected)
    checked('exactNonautonomousStages')

errors = {}
for row in data['fixtures']['integrations']:
    result = row['result']
    assert result['status'] == 'reached the requested horizon'
    assert result['endTime'] == 1.7
    reference = 1.7 / 2 - 0.25 + 1.75 * math.exp(-3.4)
    error = abs(result['history'][-1]['state'] - reference)
    errors.setdefault(row['method'], []).append(error)
    for previous, current in zip(result['history'], result['history'][1:]):
        exact, _ = exact_step(row['method'], F(previous['time']), F(previous['state']), F(current['step']))
        close(current['state'], exact, absolute=4e-15)
    checked('nonautonomousEqualHorizon')
for values in errors.values():
    assert values[2] < values[1] < values[0]

namespaces = {}
for name, example in data['examples'].items():
    output = io.StringIO()
    namespace = {}
    with contextlib.redirect_stdout(output):
        exec(compile(example['code'], f'actual-ode-{name}.py', 'exec'), namespace)
    assert output.getvalue().strip() == example['expected'].strip(), (name, output.getvalue(), example['expected'])
    namespaces[name] = namespace
    checked('exactDisplayedPrograms')

# Changed calls through the displayed helpers, not merely separate copies.
for duration in (0, 0.125, 1.25):
    for mass in (0.5, 2):
        for initial in ([2, -1], [-1, 3]):
            actual = namespaces['oscillator']['motion'](duration, mass, 0, 8, initial)
            frequency = math.sqrt(8 / mass)
            q = initial[0] * math.cos(frequency * duration) + initial[1] * math.sin(frequency * duration) / frequency
            v = -initial[0] * frequency * math.sin(frequency * duration) + initial[1] * math.cos(frequency * duration)
            np.testing.assert_allclose(actual, [q, v, (mass * v * v + 8 * q * q) / 2, 0], rtol=3e-13, atol=2e-13)
            checked('changedNativeMotion')
for duration in (0, 2**-20, 0.625, 2):
    for matrix in (np.zeros((2, 2)), np.array([[0, 2], [0, 0]]), np.diag([-0.5, 0])):
        inputs = np.array([[1, 0], [-2, 3]], dtype=float)
        transition, response = namespaces['heldInput']['held_input_matrices'](matrix, inputs, duration)
        reference = mp.expm(mp.matrix(matrix.tolist()) * duration)
        np.testing.assert_allclose(transition, np.array(reference.tolist(), dtype=float), rtol=2e-12, atol=1e-14)
        for i in range(2):
            for j in range(2):
                integral = quad(lambda t: float((mp.expm(mp.matrix(matrix.tolist()) * t) * mp.matrix(inputs.tolist()))[i, j]), 0, duration, epsabs=1e-12)[0]
                close(float(response[i, j]), integral, absolute=2e-14)
        checked('changedNativeHeldInput')
for initial in ((-2, 3), (1, -4), (0, 2)):
    coefficients = namespaces['series']['coefficients'](*initial, 24)
    # Airy basis is independent of the power-coefficient recurrence.
    basis = mp.matrix([[mp.airyai(0), mp.airybi(0)], [mp.airyai(0, 1), mp.airybi(0, 1)]])
    weights = mp.lu_solve(basis, mp.matrix(initial))
    for time in (-0.75, 0.375, 0.75):
        reference = weights[0] * mp.airyai(time) + weights[1] * mp.airybi(time)
        close(namespaces['series']['evaluate'](coefficients, time), reference, absolute=2e-13)
        checked('changedNativeSeriesAiry')
for method in ('euler', 'midpoint', 'rk4'):
    history, status = namespaces['fixedSteps']['integrate'](method, lambda t, y: t - 2 * y, 1.5, 1.7, 0.15)
    matching = next(row for row in data['fixtures']['integrations'] if row['method'] == method and row['step'] == 0.15)
    assert status == 'reached horizon'
    np.testing.assert_allclose(history, [[row['time'], row['state']] for row in matching['result']['history']], rtol=2e-13, atol=2e-14)
    checked('changedNativeStepHelper')
threshold = namespaces['events']['threshold']
result = solve_ivp(lambda t, y: -0.4 * y, (0, 10), [80.0], events=threshold, rtol=1e-10, atol=1e-12)
assert result.status == 1 and result.success
close(result.t_events[0][0], math.log(8) / 0.4, relative=2e-9)
assert result.t[-1] < 10
checked('changedNativeEvent')

result = {'checkedAt': datetime.now(timezone.utc).isoformat(), 'passed': True,
          'sources': data['sources'], 'counts': counts, 'conservedAuthorStates': data['conservedStates'],
          'comparedFields': data['comparedFields'], 'maximumAbsoluteConservationChange': data['maximumAbsoluteChange'],
          'invalidInputsRejected': data['rejected'], 'refinementErrors': errors,
          'environment': {'python': platform.python_version(), 'numpy': np.__version__, 'scipy': scipy.__version__, 'mpmath': mp.__version__}}
(directory / 'results.json').write_text(json.dumps(result, indent=2) + '\n', encoding='utf-8')
print(json.dumps(result, indent=2))

File: scripts/verify-ordinary-differential-equations-independent.mjs
import fs from 'node:fs';
import assert from 'node:assert/strict';
import { createHash } from 'node:crypto';
import * as model from '../src/learn/data/ordinary-differential-equations-models.js';
import * as original from '../scratch/ordinary-differential-equations-independent/author-sources/ordinary-differential-equations-models.js';
import { ordinaryDifferentialEquationsExamples as examples } from '../src/learn/data/ordinary-differential-equations-examples.js';

const directory = 'scratch/ordinary-differential-equations-independent';
const baseline = JSON.parse(fs.readFileSync(`${directory}/author-baseline.json`, 'utf8'));
const sources = baseline.sources.map(({ path }) => ({ path, sha256: createHash('sha256').update(fs.readFileSync(path)).digest('hex') }));
const fixtures = { cooling: [], logistic: [], matrices: [], forcing: [], stages: [], integrations: [], practice: [] };
let comparedFields = 0, maximumAbsoluteChange = 0, conservedStates = 0;
function compare(actual, expected) {
  if (typeof expected === 'number') {
    comparedFields += 1;
    const difference = Math.abs(actual - expected);
    assert(difference <= 1e-10 * Math.max(1, Math.abs(expected)), `${actual} != ${expected}`);
    maximumAbsoluteChange = Math.max(maximumAbsoluteChange, difference);
  } else if (expected && typeof expected === 'object') {
    assert.deepEqual(Object.keys(actual), Object.keys(expected));
    for (const key of Object.keys(expected)) compare(actual[key], expected[key]);
  } else assert.equal(actual, expected);
}
// Preserve author-checked inputs; changed small-number inputs have separate oracles below.
const authorFixtures = JSON.parse(fs.readFileSync('scratch/ordinary-differential-equations-verification/model-fixtures.json', 'utf8'));
const calls = {
  cooling: row => ['coolingState', [row.time, row.initial, row.rate, 10]],
  logistic: row => ['logisticState', [row.time, row.initial, row.rate, row.capacity]],
  waiting: row => ['waitingState', [row.time, row.departure]],
  matrices: row => ['exponential2', [row.matrix, row.time]],
  oscillators: row => ['oscillatorState', [row.time, row.damping, 1, row.velocity]],
  forcing: row => ['forcingTimeline', [row.time, 24, row.firstPower, row.secondPower, row.switchTime]],
  steps: row => ['odeStep', [row.method, (_time, state) => -row.rate * state, 0, 40, row.step]],
  integrations: row => ['numericalCooling', [row.method, row.step, 0.2, 5, 40]],
  amplification: row => ['coolingAmplification', [row.rate, row.step]],
  schedules: row => ['shearSchedule', [row.initial, row.first]],
};
for (const [kind, rows] of Object.entries(authorFixtures)) for (const row of rows) {
  const [name, args] = calls[kind](row);
  compare(model[name](...args), original[name](...args));
  conservedStates += 1;
}
for (const initial of [0, Number.MIN_VALUE, 1e-310, 1e-200, 2 ** -50, 3, 999999]) {
  for (const equilibrium of [0, 1e-200, 1, 1000]) for (const time of [0, 1e-200, 1e-14, 0.3, 19]) {
    fixtures.cooling.push({ initial, equilibrium, time, rate: 0.7, result: model.coolingState(time, initial, 0.7, equilibrium) });
  }
}
for (const initial of [Number.MIN_VALUE, 1e-300, 0.25, 7, 900]) for (const capacity of [0.01, 7, 91]) for (const time of [0.25, 0.75, 7]) {
  fixtures.logistic.push({ initial, capacity, time, rate: 0.3, result: model.logisticState(time, initial, 0.3, capacity) });
}
for (const diagonal of [-3, 0, 1]) for (const coupling of [-7, 0, 9]) for (const time of [-0.75, 0, 0.125, 1.75]) {
  const matrix = [[diagonal, coupling], [0, diagonal]];
  fixtures.matrices.push({ matrix, time, result: model.exponential2(matrix, time) });
}
for (const sign of [-1, 1]) for (const near of [0, 2 ** -35, -(2 ** -35)]) for (const time of [0.03125, 0.625, 3]) {
  const matrix = [[-2, sign * (4 + near)], [-sign, 2]];
  fixtures.matrices.push({ matrix, time, result: model.exponential2(matrix, time) });
}
for (const switchTime of [0, 0.75, 2.25, 6]) for (const time of [0.25, 0.75, 2.25, 5.75]) {
  fixtures.forcing.push({ switchTime, time, initial: 13, firstPower: 35, secondPower: 5,
    result: model.forcingTimeline(time, 13, 35, 5, switchTime) });
}
for (const method of ['euler', 'midpoint', 'rk4']) for (const time of [0, 0.375, 1.25]) for (const step of [0.0625, 0.25, 0.875]) {
  const state = 1.5, derivative = (t, y) => t - 2 * y;
  fixtures.stages.push({ method, time, step, state, result: model.odeStep(method, derivative, time, state, step) });
}
for (const method of ['euler', 'midpoint', 'rk4']) for (const step of [0.3, 0.15, 0.075]) {
  fixtures.integrations.push({ method, step, result: model.integrateScalar({ method, derivative: (t, y) => t - 2 * y, initial: 1.5, endTime: 1.7, step }) });
}
for (const endpoint of ['pi', 'half-pi']) for (const target of [-3, 0, 0.25, 2]) {
  const result = model.bvpFamily(endpoint, target, -1.5);
  assert.equal(result.classification, endpoint === 'half-pi' ? 'one solution' : target === 0 ? 'infinitely many solutions' : 'no solution');
}
const withInherited = (array, key, value) => {
  const prototype = Object.create(Array.prototype);
  prototype[key] = value;
  Object.setPrototypeOf(array, prototype);
  return array;
};
const badVectors = [[1, ,], withInherited([1, ,], 1, 2), [NaN, 1], [1, Infinity], [true, 1]];
let rejected = 0;
for (const vector of badVectors) for (const action of [
  () => model.applyOdeMatrix([[1, 0], [0, 1]], vector),
  () => model.linearSystemState([[1, 0], [0, 1]], vector, 1),
  () => model.shearSchedule(vector),
  () => model.exponential2([vector, [0, 1]], 1),
]) { assert.throws(action); rejected += 1; }
for (const action of [
  () => model.exponential2(withInherited([[1, 0], ,], 1, [0, 1]), 1),
  () => model.coolingState(1e-320, 0, 1e-5, 1e6),
  () => model.odeStep('rk4', () => Infinity, 0, 1, 0.1),
  () => model.odeStep('rk4', () => 1, 1e6, 1, Number.MIN_VALUE),
  () => model.bvpFamily('constructor', 0),
]) { assert.throws(action); rejected += 1; }
assert.equal(model.coolingState(0, 1e-200, 1, 1).state, 1e-200);
assert.equal(model.coolingState(1e-200, 1e-200, 1, 1).state, 2e-200);
const budget = model.integrateScalar({ derivative: () => 1, initial: 0, endTime: 10, step: 0.3, maximumSteps: 7 });
assert.match(budget.status, /exhausted/);
assert.equal(budget.steps, 7);
const stopped = model.integrateScalar({ derivative: () => NaN, initial: 0, endTime: 1, step: 0.1 });
assert.match(stopped.status, /nonfinite/);
assert.equal(stopped.steps, 0);
fs.writeFileSync(`${directory}/fixtures.json`, JSON.stringify({ at: new Date().toISOString(), sources, conservedStates, comparedFields, maximumAbsoluteChange, rejected, fixtures, examples }, null, 2));
console.log(JSON.stringify({ conservedStates, comparedFields, maximumAbsoluteChange, rejected, changedCases: Object.fromEntries(Object.entries(fixtures).map(([name, rows]) => [name, rows.length])) }));
